### PR TITLE
Replace CScriptID and CKeyID in CTxDestination with dedicated types

### PR DESCRIPTION
**Pull request by instagibbs**
The current usage seems to be an overloading of meanings. `CScriptID` is used in the wallet as a lookup key, as well as a destination, and `CKeyID` likewise. Instead, have all destinations be dedicated types.

New types:
`CScriptID`->`ScriptHash`
`CKeyID`->`PKHash`

### CODE REVIEW

File: src/script/standard.h [hunk 1]
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
...
 
+struct PKHash : public uint160
+{
+    PKHash() : uint160() {}
+    explicit PKHash(const uint160& hash) : uint160(hash) {}
+    explicit PKHash(const CPubKey& pubkey);
+    using uint160::uint160;
+};
+
+struct SHash : public uint160
```
Comment (laanwj) [inline src/script/standard.h]: Concept ACK—I dislike the name `SHash` though, it sounds too general. I'd prefer a longer name that makes it more clear what is being hashed, maybe `ScriptHash`?

Comment (instagibbs) [reply src/script/standard.h]: done

**Comment by sipa** [comment]
Concept ACK. I believe @ryanofsky suggested something like this before.

File: src/script/standard.h [hunk 2]
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
...
+    explicit PKHash(const uint160& hash) : uint160(hash) {}
+    explicit PKHash(const CPubKey& pubkey);
+    using uint160::uint160;
+};
+
+struct ScriptHash : public uint160
+{
+    ScriptHash() : uint160() {}
+    explicit ScriptHash(const uint160& hash) : uint160(hash) {}
+    explicit ScriptHash(const CScript& script);
```
Comment (practicalswift) [inline src/script/standard.h]: Nit: Make sure parameter names match between declaration and definition :-)

**Comment by instagibbs** [comment]
Not sure how to read appveyor error tbh

**Comment by MarcoFalke** [comment]
The appveyor sync failure was fixed earlier this week

**Comment by Sjors** [comment]
Concept ACK. Maybe enumerate the new types in the PR description?

Shouldn't `CKeyID GetID()` in `pubkey.h` also return a  `PKHash `? That breaks a few other things from the looks of it, but otherwise it seems inconsistent.

File: src/script/standard.h [hunk 3]
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
     friend bool operator<(const CNoDestination &a, const CNoDestination &b) { return true; }
 };
 
+struct PKHash : public uint160
+{
+    PKHash() : uint160() {}
+    explicit PKHash(const uint160& hash) : uint160(hash) {}
+    explicit PKHash(const CPubKey& pubkey);
```
Comment (Sjors) [inline src/script/standard.h]: Inline definition might be more clear but I don't want to upset any strong c++ preferences out there.

Comment (instagibbs) [reply src/script/standard.h]: It's a mixed bag here(definition for other constructors are in .cpp), I'll leave as-is unless someone else chimes in

**Comment by stevenroose** [comment]
utACK 3330398ab5df6f67859307339c0bb71c6b687220

**Comment by instagibbs** [comment]
rebased

**Comment by instagibbs** [comment]
rebased and added description of new destination types in OP @Sjors

**Review by ryanofsky — APPROVED** [review]
Conditional utACK 51482131c645ad565ad032a57bd105ede65f9b86 if fixes are added for travis failures. I'm guessing this will just require rebasing and adding some more CKeyID/CScriptID replacements.

This change was also suggested a while ago in: https://github.com/bitcoin/bitcoin/pull/11403#discussion_r155821355 and I think it could be simplified a little by taking the other suggestion in that comment and making ScriptHash inherit from CScriptID and PKHash inherit from CKeyID. This would avoid the need to add explicit conversions several places in this PR. I think it's important for conversions from CScriptID to ScriptHash and CKeyID to PKHash to be explicit, but it seems fine for conversions in the opposite direction to be implicit.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3768,7 +3768,7 @@ void CWallet::GetKeyBirthTimes(interfaces::Chain::Lock& locked_chain, std::map<C
     // get birth times for keys with metadata
     for (const auto& entry : mapKeyMetadata) {
         if (entry.second.nCreateTime) {
-            mapKeyBirth[entry.first] = entry.second.nCreateTime;
+            mapKeyBirth[PKHash(entry.first)] = entry.second.nCreateTime;
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/wallet/wallet.cpp]: done

**Comment by instagibbs** [comment]
fixed a single dangling test case that was expecting a `PKHash`, and added a commit to change GetKeyBirthTimes to returning `CKeyID` instead of `CTxDestination`.

**Review by ryanofsky — APPROVED** [review]
utACK ddc21f2f2360a3d7e7acdddd4478822d47c44f7a. Changes since last review are fixing test case and switching GetKeyBirthTimes away from CTxDestination.

**Comment by Sjors** [comment]
tACK ddc21f2. Compiled, opened a wallet and ran test suite and bench on macOS 10.14.4. Not sure what else to test here.

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -571,7 +571,7 @@ static UniValue decodescript(const JSONRPCRequest& request)
     if (type.isStr() && type.get_str() != "scripthash") {
         // P2SH cannot be wrapped in a P2SH. If this script is already a P2SH,
         // don't return the address for a P2SH of the P2SH.
-        r.pushKV("p2sh", EncodeDestination(CScriptID(script)));
+        r.pushKV("p2sh", EncodeDestination(ScriptHash(CScriptID(script))));
```
Comment (meshcollider) [inline src/rpc/rawtransaction.cpp]: `ScriptHash(script)`? Same on line 601

File: src/test/script_standard_tests.cpp [hunk 6]
```diff
@@ -490,8 +490,8 @@ BOOST_AUTO_TEST_CASE(script_standard_IsMine)
     {
         CBasicKeyStore keystore;
 
-        CScript redeemscript = GetScriptForDestination(pubkeys[0].GetID());
-        CScript witnessscript = GetScriptForDestination(CScriptID(redeemscript));
+        CScript redeemscript = GetScriptForDestination(ScriptHash(PKHash(pubkeys[0])));
```
Comment (meshcollider) [inline src/test/script_standard_tests.cpp]: Why is there `ScriptHash` here?

File: src/test/script_standard_tests.cpp [hunk 7]
```diff
@@ -598,7 +598,7 @@ BOOST_AUTO_TEST_CASE(script_standard_IsMine)
         BOOST_CHECK(keystore.AddKey(keys[1]));
 
         CScript redeemScript = GetScriptForMultisig(2, {uncompressedPubkey, pubkeys[1]});
-        scriptPubKey = GetScriptForDestination(CScriptID(redeemScript));
+        scriptPubKey = GetScriptForDestination(ScriptHash(CScriptID(redeemScript)));
```
Comment (meshcollider) [inline src/test/script_standard_tests.cpp]: `ScriptHash(redeemScript)` again?

File: src/test/script_standard_tests.cpp (hunk 6 above)
Comment (instagibbs) [reply src/test/script_standard_tests.cpp]: hmm seems I broke this test case, thanks

File: src/test/script_standard_tests.cpp (hunk 7 above)
Comment (instagibbs) [reply src/test/script_standard_tests.cpp]: fixed

File: src/rpc/rawtransaction.cpp (hunk 5 above)
Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: fixed

**Comment by instagibbs** [comment]
all comments addressed

**Review by ryanofsky — APPROVED** [review]
utACK 78e407ad0c26190a22de1bc8ed900164a44a36c3. Only changes are removing extra CScriptID()s and fixing the test case.

**Comment by Sjors** [comment]
utACK 78e407a

**Review by Sjors — APPROVED** [review]
utACK 78e407a

File: src/test/script_standard_tests.cpp (hunk 7 above)
Comment (Sjors) [reply src/test/script_standard_tests.cpp]: A followup PR that would make this particular mistake fail the test would be nice though.

**Review by meshcollider — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/15452/commits/78e407ad0c26190a22de1bc8ed900164a44a36c3

By removal of the `// TODO: include scripts in GetKeyBirthTimes() output instead of separate` comment, I assume you think that shouldn't be done, rather than because you implemented it.

**Comment by instagibbs** [comment]
@meshcollider payoff was very low, and descriptor wallets would have changed whatever someone was going to do

**Pull request merged**